src/core/mesh/qgsmeshdataprovidertemporalcapabilities.cpp

```diff
@@ -0,0 +1,175 @@
+/***************************************************************************
+                         qgsmeshdataprovidertemporalcapabilities.cpp
+                         -----------------------
+    begin                : March 2020
+    copyright            : (C) 2020 by Vincent
+    email                : zilolv at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmeshdataprovidertemporalcapabilities.h"
+
+
+QgsMeshDataProviderTemporalCapabilities::QgsMeshDataProviderTemporalCapabilities(): QgsDataProviderTemporalCapabilities()
+{}
+
+QgsMeshDatasetIndex QgsMeshDataProviderTemporalCapabilities::datasetIndexTimeInMilliseconds( int group, qint64 timeSinceGlobalReference ) const
+{
+  const QDateTime &groupReference = mGroupsReferenceDateTime[group];
+
+  const qint64 timeSinceGroupeReference =
+    timeSinceGlobalReference - mGlobalReferenceDateTime.msecsTo( groupReference );
+  const QList<qint64> &datasetTimes = mDatasetTimeSinceGroupReference[group];
+  // No dataset
+  if ( datasetTimes.isEmpty() )
+    return QgsMeshDatasetIndex();
+
+  // If requested time is before the timestamp of the first dataset, return the first dataset
```

Comment:
With mesh layer, some datasets (for example, terrain elevation) are not really temporal and there is no reason to not display them for particular time step. But, MDAL doesn't provide way to distinguish "static" datasets and "dynamic" datasets. I used this workaround to always display such datasets, but, sure, it is not very clean... We can add the condition that the dataset group has only one dataset but it is not completely satisfying...

Reply:
I would argue that this situation should be handled by users disabling the temporal option for those layers - that has the effect that the layer is always shown.

One option that is missing here is an equivalent of the raster temporal properties "fixed time range" mode, where a user can manually set a particular time range for that layer to shown within. It's in place to allow the animation framework to work with non-temporal-aware data sources (e.g. geotiff) where a user may have a set of multiple layers representing different time periods, and wants the animation to automatically switch them on or off if they fall within this fixed time range.

I think the mesh layers could also benefit from this option -- it's a "last resort" fallback for users allowing maximum flexibility for animations.

Reply:
Set the whole layer static could be a solution for the user, but for mesh layer, there is not only one dataset group, we can have a terrain elevation dataset and many other dynamic dataset groups. I think also about the min/max dataset group...
More over, the problem is the same when the request time is greater than the time extent of the datasets. That is like this because the request of dataset is done with only one relative time value.
I wonder if it will not be better to request dataset with interval of relative time, and add an option to allow the display with the extreme dataset when the request time interval is before or after the time extent of the datasets?

Reply:
discussed with Vincent: 

1. implement `MDAL_EXPORT bool MDAL_G_isTemporal( DatasetGroupH group );`
2. add `isTemporal()` to QgsMeshDatasetGroup class
3. check the code that if group is not temporal, display always regardless of the time settings

Reply:
as for "fixed time range" mode, I would propose to have it available (optionally) only for static mesh layers. for dynamic I do not see much sense

Reply:
Ok, I'll defer to your superior knowledge of mesh internals here! :+1: